src/planner/constify_now.c

```diff
@@ -0,0 +1,202 @@
+ * To enable this behaviour we convert `column > now()` expressions
+ * into `column > const AND column > now()`. Assuming that times
+ * always moves forward this is save even for prepared statements.
+ *
+ * We consider the following expressions valid for this optimization:
+ * - Var > now()
+ * - Var >= now()
+ * - Var > now() - Interval
+ * - Var > now() + Interval
+ * - Var >= now() - Interval
+ * - Var >= now() + Interval
+ *
+ * Additionally Interval needs to be Const and not contain day or month
+ * components as those would be affected by timezone.
+ */
+static const Dimension *
+get_hypertable_dimension(Oid relid)
+{
+	Hypertable *ht = ts_planner_get_hypertable(relid, CACHE_FLAG_CHECK);
+	if (!ht)
+		return NULL;
+	return hyperspace_get_open_dimension(ht->space, 0);
+}
+
+static bool
+is_valid_now_expr(OpExpr *op, List *rtable)
+{
+	/* Var > or Var >= */
+	if ((op->opfuncid != F_TIMESTAMPTZ_GT && op->opfuncid != F_TIMESTAMPTZ_GE) ||
+		!IsA(linitial(op->args), Var))
+		return false;
+
+	/*
+	 * Check that the constraint is actually on a partitioning
+	 * column. We only check for match on first open dimension
+	 * because that will be the time column.
+	 */
+	Var *var = linitial_node(Var, op->args);
+	Assert(var->varno <= list_length(rtable));
+	RangeTblEntry *rte = list_nth(rtable, var->varno - 1);
+
+	const Dimension *dim = get_hypertable_dimension(rte->relid);
+	if (!dim || dim->fd.column_type != TIMESTAMPTZOID || dim->column_attno != var->varattno)
+		return false;
+
+	/* Var > now() or Var >= now() */
+	if (IsA(lsecond(op->args), FuncExpr) && lsecond_node(FuncExpr, op->args)->funcid == F_NOW)
+		return true;
+
+	if (!IsA(lsecond(op->args), OpExpr))
+		return false;
+
+	/* Var >|>= now() +|- Const */
+	OpExpr *op_inner = lsecond_node(OpExpr, op->args);
+	if ((op_inner->opfuncid != F_TIMESTAMPTZ_MI_INTERVAL &&
+		 op_inner->opfuncid != F_TIMESTAMPTZ_PL_INTERVAL) ||
+		!IsA(linitial(op_inner->args), FuncExpr) ||
+		linitial_node(FuncExpr, op_inner->args)->funcid != F_NOW ||
+		!IsA(lsecond(op_inner->args), Const))
+		return false;
+
+	/*
+	 * The consttype check should not be necessary since the
+	 * operators we whitelist above already mandates it.
+	 */
+	Const *c = lsecond_node(Const, op_inner->args);
+	Assert(c->consttype == INTERVALOID);
+	if (c->constisnull || c->consttype != INTERVALOID)
+		return false;
+
+	Interval *offset = DatumGetIntervalP(c->constvalue);
+	/*
+	 * We don't consider day or month intervals safe here as
+	 * they are affected by timezones and therefore not
+	 * safe to evaluate during planning.
+	 */
+	if (offset->day != 0 || offset->month != 0)
+		return false;
+
+	return true;
+}
+
+static Const *
+make_now_const()
+{
+	return makeConst(TIMESTAMPTZOID,
+					 -1,
+					 InvalidOid,
+					 sizeof(TimestampTz),
+#ifdef TS_DEBUG
+					 ts_get_mock_time_or_current_time(),
+#else
+					 TimestampTzGetDatum(GetCurrentTransactionStartTimestamp()),
```

Comment:
Hmm, we use transaction start time here?

Reply:
Thats what now uses:
```
Datum
now(PG_FUNCTION_ARGS)
{
  PG_RETURN_TIMESTAMPTZ(GetCurrentTransactionStartTimestamp());
}
```